apps/snitch_core/lib/core/data/schema/card.ex

```diff
@@ -1,47 +1,84 @@
 defmodule Snitch.Data.Schema.Card do
   use Snitch.Data.Schema
 
-  alias Snitch.Data.Schema.{CardPayment, Address, User}
+  alias Snitch.Data.Schema.{CardPayment, User}
 
   @type t :: %__MODULE__{}
 
   schema "snitch_cards" do
-    field(:last_digits, :string)
     field(:name_on_card, :string)
     field(:year, :integer)
     field(:month, :integer)
-    field(:card_name, :string)
     field(:brand, :string)
+    field(:is_disabled, :boolean, default: false)
+    field(:number, :string)
+
+    field(:card_name, :string, virtual: true)
     field(:first_name, :string, virtual: true)
     field(:last_name, :string, virtual: true)
+    field(:last_digits, :string, virtual: true)
 
-    belongs_to(:address, Address)
     belongs_to(:user, User)
 
     has_many(:card_payments, CardPayment)
     timestamps()
   end
 
-  @required_fields ~w(last_digits name_on_card brand user_id month year)a
-  @optional_fields ~w(card_name address_id)a
-
+  @required_fields ~w(name_on_card brand user_id month year number)a
+  @update_fields ~w(is_disabled)a
   @spec changeset(__MODULE__.t(), map, :create | :update) :: Ecto.Changeset.t()
   def changeset(card, params, action)
 
   def changeset(card, params, :create) do
     %{year: current_year, month: current_month} = DateTime.utc_now()
 
+    # for now only for VISA cards.
+
     card
-    |> cast(params, @required_fields ++ @optional_fields)
+    |> cast(params, @required_fields)
     |> validate_required(@required_fields)
+    |> build_card
     |> validate_number(
       :month,
       less_than_or_equal_to: 12,
       greater_than_or_equal_to: current_month
     )
     |> validate_number(:year, greater_than_or_equal_to: current_year)
-    |> validate_format(:last_digits, ~r/^\d{4}$/)
-    |> foreign_key_constraint(:address_id)
+    |> validate_format(:number, ~r/^4[0-9]{12}(?:[0-9]{3})?$/)
     |> foreign_key_constraint(:user_id)
   end
+
+  def changeset(card, params, :update) do
+    cast(card, params, @update_fields)
+  end
+
+  def build_card(card) do
+    card
+    |> split_name(Map.has_key?(card.changes, :name_on_card))
+    |> extract_last_digits(
+      Map.has_key?(card.changes, :card_number),
+      Map.has_key?(card.changes, :brand)
+    )
+  end
+
+  defp split_name(card, true) do
```

Comment:
This is a lossy function! `split_name("A. K. Jadhav")` will simply throw away the middle `K.`. IMO opinion we should take a simple approach:
1. If `:name_on_card` is set, we will leave `:first_name` and `:last_name` empty.
2. If `:name_on_card` is empty and `:first_name`, `:last_name` are set, we use them to build `:name_on_card`.
3. If all name fields are empty, we mark changeset invalid.